src/riemannian_diff.jl

```diff
@@ -9,97 +9,70 @@ abstract type AbstractRiemannianDiffBackend end
+Compute the Riemannian Jacobian of a map of type `mt` represented by function `f`
+at point `p` using the given backend.
 """
-r_jacobian(::AbstractMap, ::Any, ::AbstractRiemannianDiffBackend)
+jacobian(::Any, ::AbstractMap, ::Any, ::AbstractRiemannianDiffBackend)
 
-function r_differential(f::AbstractCurve, t, backend::AbstractRiemannianDiffBackend)
-    return error(
-        "r_differential not implemented for curve $(typeof(f)), point $(typeof(t)) and " *
-        "backend $(typeof(backend))",
-    )
+function differential!(f::Any, c::Curve, X, t, backend::AbstractRiemannianDiffBackend)
+    return copyto!(X, differential(f, c, t, backend))
 end
 
-function r_differential!(f::AbstractCurve, X, t, backend::AbstractRiemannianDiffBackend)
-    return copyto!(X, r_differential(f, t, backend))
+function gradient!(f, ft::RealField, X, p, backend::AbstractRiemannianDiffBackend)
+    return copyto!(X, gradient(f, ft, p, backend))
 end
 
-function r_gradient(f::AbstractRealField, p, backend::AbstractRiemannianDiffBackend)
-    return error(
-        "r_gradient not implemented for field $(typeof(f)), point $(typeof(p)) and " *
-        "backend $(typeof(backend))",
-    )
-end
-
-function r_gradient!(f::AbstractRealField, X, p, backend::AbstractRiemannianDiffBackend)
-    return copyto!(X, r_gradient(f, p, backend))
-end
-
-function r_hessian(f::AbstractRealField, p, backend::AbstractRiemannianDiffBackend)
-    return error(
-        "r_hessian not implemented for field $(typeof(f)), point $(typeof(p)) and " *
-        "backend $(typeof(backend))",
-    )
-end
+"""
+    hessian_vector_product(f, ft::RealField, p, X, backend::AbstractRiemannianDiffBackend)
 
-function r_hessian_vector_product(
-    f::AbstractRealField,
-    p,
-    X,
-    backend::AbstractRiemannianDiffBackend,
-)
-    return error(
-        "r_hessian_vector_product not implemented for field $(typeof(f)), point $(typeof(p)), vector $(typeof(X)) and " *
-        "backend $(typeof(backend))",
-    )
-end
 
-function r_jacobian(f::AbstractMap, p, backend::AbstractRiemannianDiffBackend)
-    return error(
-        "r_jacobian not implemented for map $(typeof(f)), point $(typeof(p)) and " *
-        "backend $(typeof(backend))",
-    )
-end
+Compute the product of Riemannian Hessian of a real field of type `rf` represented by
+function `f` at point `p` and the tangent vector `X` at point `p` using the given backend.
+"""
+hessian_vector_product(f, ft::RealField, p, X, backend::AbstractRiemannianDiffBackend)
 
-r_differential(f::AbstractCurve, p) = r_differential(f, p, rdiff_backend())
+differential(f, c::Curve, p) = differential(f, c, p, rdiff_backend())
 
-r_differential!(f::AbstractCurve, X, p) = r_differential!(f, X, p, rdiff_backend())
+differential!(f, c::Curve, X, p) = differential!(f, c, X, p, rdiff_backend())
 
-r_gradient(f::AbstractRealField, p) = r_gradient(f, p, rdiff_backend())
+gradient(f, rf::RealField, p) = gradient(f, rf, p, rdiff_backend())
```

Comment:
I would still prefer seeing the manifold here, since there might be gradients that require information about the manifold or for conversion a basis in TpM, then it would be neat to have `M` here. Same for the differential and the hessian and the hessian vector product. For those two I think you mean that the hessian is `H(p)` while the product is `H(p)[X]`?

Reply:
`rf` contains a manifold, and it can be easily extended to also contain other information. Maybe I should introduce `AbstractRealField` here though. This way it's straightforward to introduce additional information but putting it in `rf` instead of adding additional arguments (which will mess with AD backend information or, if keyword arguments, can't be dispatched over).

> For those two I think you mean that the hessian is `H(p)` while the product is `H(p)[X]`?

Yes, and more specifically, Hessian in my code is the matrix of Hessian coefficients in the given basis and Hessian-vector product is the tangent vector that's the value of Hessian on the given tangent vector (not its coefficients in a basis).

I'll have to think how to incorporate other representations of the Hessian but most likely it would be specified by `AbstractRiemannianDiffBackend`. And I'm going to separate `AbstractRiemannianDiffBackend` into different backends for gradients, hessians and differentials.

Reply:
Thanks for the clarification, you'Re right, the manifold is withing `rf`, I was not careful enough trying to understand that.

For the hessian, we could nearly combine those right?
A `hessian(f,p)`, where `f` is a `AbstractRealField` has both manifolds, so one can take their `DefaultBases` as a matrix representation but it should maybe also be possible to specify both basis, i.e. `hessian(f,p,b1,b2)` - where `b1,b2` are `AbstractBase`s. For the hessian applied to a vector we have `hessian(f,p,X)`. What do you think?

Reply:
> For the hessian, we could nearly combine those right?

Yes, it (`f` and `ft`) was even combined before. I'm not quite sure which version I prefer.

> A `hessian(f,p)`, where `f` is a `AbstractRealField` has both manifolds

What manifolds? I can see only one for Hessians.

> For the hessian applied to a vector we have `hessian(f,p,X)`

I would prefer to have different functions for coefficients of Hessians in bases and values of Hessians on tangent vectors.

Reply:
Ah right, for the hessian it's only one.

I am not a fan of the currect hessian-vector-name, that's why I am looking for a new name – and I think it could be combined into one name, why should it not?

Reply:
Making both variants methods of one function will cause ambiguity issues and, from my previous experience, I prefer to avoid method ambiguity rather than solving it.

I agree that `hessian_vector_product` is not a great name. Maybe the discussion from https://github.com/JuliaManifolds/ManifoldsBase.jl/issues/45 would be relevant here?

Reply:
Ah the ambiguities yes, that might be a problem.

Yes that reference is relevant here but from your current approach it solves my question therein – we just have to find a nice name :)